AVX-512 optimizations.

Sopel97 opened the pull request:
This branch has some things to try and benchmark. Since there are no AVX512 workers on fishtest this has to proceed somewhat differently but maybe we'll reach consensus.

Changes:
- AVX512 code for smaller affine transforms that works over 2 columns at a time.
- The above unrolled by 8 times instead of 4 times, with more parallel horizontal adds (`m512_hadd256x16`).
- (behind #if 0, because slightly worse in limited tests. Requires further investigation) Unrolled AVX512 affine transform loop over output dimensions from 4 to 8 while doing more parallel horizontal adds (`m512_haddx8`).

- AVX512 code for feature transformer `Transform`

I'm fairly certain that this code in this PR is better than master as tests by ipmanchess indicate about 1.5% speed improvement. The thing I'm not sure about is the last bullet point - the unrolling of the normal AVX512 loop - which turned out inconclusive. Now this was tested without a profile build as I can't make one. The results could be different with a profile build, and possibly differ by hardware slightly.

vondele commented:
on aws t3.xlarge Platinum 8259CL CPU @ 2.50GHz
```
Result of  10 runs
==================
base (...kfish.master) =    1795390  +/- 3762
test (...ckfish.patch) =    1828041  +/- 2566
diff                   =     +32651  +/- 2849

speedup        = +0.0182
P(speedup > 0) =  1.0000
```

vondele commented:
if I remove the '#if 0' (and fix a small compilation issue), the benefit is clearly not as big (0.3% instead 1.8%).

Sopel97 commented:
That aligns with ipmanchess' results. Thanks

Sopel97 commented:
I've rebased on the better gcc bug workaround, removed the 8x unrolled avx512 implementation for the large layer. So this now should be the best I can come up in this area.

mstembera commented:
For clarity it would be much nicer to commit the stack array align stuff separately from this.

mstembera commented:
I am wondering why in master m512_haddx4 is implemented using the unpacklo unpackhi intrinsics but m256_haddx4 is implemented using the hadd intrinsics?  It seems we should pick the same scheme for both as a base comparison.

Sopel97 commented:
That's just because AVX512 doesn't have hadd.

vondele commented:
@Sopel97 once you think this is ready, please submit a non-regression test on fishtest, to exclude this would be slower on avx2. I seem to locally observe a small ( ~0.5%) slowdown, but it could be in the noise.

Sopel97 commented:
After some tinkering I've reverted the last change, it seems that gcc doesn't quite like it and clang generates almost identical code for both. Now the AVX2 code is not affected, if you think I should still run non-regression I can do that later.

vondele commented:
since it is now only affecting avx512 there is need to run a non-regression test. I'll repeat the speed measurement before merging.

vondele commented:
can squash in a single commit with an appropriate commit message?

mstembera commented on src/nnue/nnue_feature_transformer.h:
```diff
@@ -154,7 +160,18 @@ namespace Eval::NNUE {
+        auto out = reinterpret_cast<__m512i*>(&output[offset]);
+        for (IndexType j = 0; j < kNumChunks; ++j) {
+          __m512i sum0 = _mm512_load_si512(
+              &reinterpret_cast<const __m512i*>(accumulation[perspectives[p]][0])[j * 2 + 0]);
```
The indentations here are inconsistent.

Sopel97 replied:
I changed the indentation of like 171, don't really know whether it should be 2 or 4 spaces here because it differs from line to line in this file

mstembera replied:
Line 167 and 169 should be the same but aren't.  I'm not sure about 2 or 4 spaces for the same reason as you.

gvreuls commented:
> This branch has some things to try and benchmark. Since there are no AVX512 workers on fishtest this has to proceed somewhat differently but maybe we'll reach consensus.

There are AVX-512 capable workers on fishtest but the AVX-512 build on fishtest is disabled because it's "too slow" (which isn't true for my AVX-512 system). Will this influence the decision not to use AVX-512 builds on fishtest?

vondele commented:
@gvreuls would be interesting if you could benchmark (single-threaded and multithreaded) this patch on your system relative to avx2. I'm not opposed to enabling avx512 on fishtest if we figure out in which case it is a win relative to avx2.

Sopel97 commented:
@vondele from ipmanchess I got the following for the previous PR https://cdn.discordapp.com/attachments/771338720028393472/772365509307924490/Stockfish_avx2-avx512_pgo_benches.PNG

gvreuls commented:
@vondele I'm not at home right now, it will take a few hours until I can run the tests.

gvreuls commented:
Here are my test results (the average of 5 bench runs). Please note that om my system AVX512 runs at the same clock speed as AVX2 (and I strongly suspect this is the case on ipmanchess' system too), I can rerun the tests with reduced AVX512 clock speed if that's deemed necessary.
```
Master AVX2 1 thread:
 Performance counter stats for 'system wide' (5 runs):

     7.414.924.848      cycles:u                                                      ( +-  0,83% )
    12.386.333.875      instructions:u            #    1,67  insn per cycle           ( +-  0,52% )

            1,7229 +- 0,0193 seconds time elapsed  ( +-  1,12% )
```
```
Sopel AVX512 1 thead:
 Performance counter stats for 'system wide' (5 runs):

     6.951.475.653      cycles:u                                                      ( +-  0,64% )
    11.342.953.826      instructions:u            #    1,63  insn per cycle           ( +-  0,38% )

            1,6052 +- 0,0161 seconds time elapsed  ( +-  1,00% )
```
```
Master AVX2 16 threads:
 Performance counter stats for 'system wide' (5 runs):

    51.001.185.587      cycles:u                                                      ( +-  1,16% )
    51.672.129.967      instructions:u            #    1,01  insn per cycle           ( +-  1,11% )

            1,0090 +- 0,0123 seconds time elapsed  ( +-  1,22% )
```
```
Sopel AVX512 16 threads:
 Performance counter stats for 'system wide' (5 runs):

    47.428.267.772      cycles:u                                                      ( +-  2,39% )
    46.959.286.438      instructions:u            #    0,99  insn per cycle           ( +-  2,44% )

            0,9363 +- 0,0252 seconds time elapsed  ( +-  2,69% )
```

vondele commented:
I also have rerun the benchmark and see 1.8% speedup for the patch. avx512 vs avx2 is 9.5% faster. 

I will merge this patch in the next round, and we should consider using avx512 on fishtest, but this can be dealt with separately.

vondele commented:
@gvreuls fishtest now using avx512 : 

gvreuls-16cores-5 | Linux 5.8.17-200.fc32.x86_64      ARCH=64bit AVX512 BMI2 AVX2 SSE41 SSSE3 SSE2 POPCNT

gvreuls commented:
@vondele Yes I noticed from https://github.com/glinscott/fishtest/issues/782#issuecomment-723506774 thanks! From what I've gathered so far pemo and Spptr also have a AVX512 systems.
